ZOOKEEPER-3173: Quorum TLS - support PEM trust/key stores

ivmaykov opened the pull request:
Add support for loading key and trust stores from PEM files.

Unfortunately, this PR includes 2 JIRAs, because it was quite difficult
to untangle the two features as they were developed at the same time
originally:

- ZOOKEEPER-3173: Quorum TLS - support PEM trust/key stores
- ZOOKEEPER-3175: Quorum TLS - test improvements

## Added support for PEM formatted key stores and trust stores
- key store and trust store files can now be in PEM format as well as JKS.
- Added config properties to tell ZK what type of trust/key store to load:
- `zookeeper.ssl.keyStore.type` and `zookeeper.ssl.trustStore.type` for ClientX509Util
- `zookeeper.ssl.quorum.keyStore.type` and `zookeeper.ssl.quorum.trustStore.type` for QuorumX509Util
- store type properties could have the values "JKS", "PEM", or not set
- leaving the type properties unset will cause auto-detection of the store type based on the file extension (".jks" or ".pem")

## Added test utilities for easily creating X509 certs and using them in unit tests
- added new class `X509TestContext` and its friend, `X509TestHelpers`
- rewrote some existing unit tests to use these classes, and added new tests that use them
- some existing tests (i.e. `QuorumSSLTest`) should probably be ported to use this as well, haven't got around to it yet

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2500/

ivmaykov commented:
The test failures don't look to be related to my changes. @anmolnar @hanm is there a way to trigger a Jenkins build re-run?

eolivelli commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2514/

ivmaykov commented:
@eolivelli should the copied code be put into the zookeeper-contrib subproject? Or can it still live in zookeeper-server? I could also use help with wording the message in the NOTICE file. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2525/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2521/

ivmaykov commented:
I should mention that this code has been internally reviewed at Facebook, has been landed on our internal fork, and has been running in production for weeks.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -221,15 +279,45 @@ public SSLContext createSSLContext(ZKConfig config) throws SSLContextException {
+                keyStoreType = detectStoreFileTypeFromFileExtension(keyStoreFile);
+            }
+            KeyStore ks;
+            switch (keyStoreType) {
```
Using switch cases to employ different implementations for the same purpose is usually a code smell to me. What do you think of implementing an abstract class or an interface to gather all keystore reader implementations and create child classes for PEM and JKS?

In which case you can,
1. pass only the interface KeystoreLoader here,
2. avoid PemReader being implemented with all static methods,
3. easily mock the impl in unit tests

eolivelli replied:
@anmolnar 
IMHO in this specific case the switch can be acceptable.
With a fully generic solution maybe we should also add a generic way to detect the type from the filename or file contents (so that we can support new file types in the future without touching this code).

Anyway any future proof implementation is always welcome

anmolnar replied:
That factory logic is already implemented in `detectStoreFileTypeFromFileExtension()` and which (I believe) should go into a static factory method or factory class.

Another potential weakness of such switch-case implementations is what you mentioned in terms of adding new container types: one has to implement the new type in each of every switch case related to this type. In this case it's only 2 which I think already quite error-prone, but there could be others later.

ivmaykov replied:
@anmolnar @eolivelli There are pros and cons to either approach. We only do the `switch (StoreFileType)` in two places, both of which are in the same file, which I feel is pretty well localized. If/when someone adds support for another key store type later (say, PKCS12), as long as they add decent tests in the same commit, I don't think there is much risk of forgetting to update one of the two switch statements. So, creating interfaces and implementing them in several classes kind of feels like over-engineering and will make it hard to see all the logic at once since it will be spread out across 7 files (assuming interface + 2 implementations for each of KeyStoreLoader / TrustStoreLoader) instead of 1 file.

But it would make mocking easier and adding more store types later would be potentially simplified, plus we could test the loaders in isolation, so I see pros and cons to both approaches. We could also land it like this in the interests of getting a working implementation out, and clean it up in a later PR. You guys let me know what you would prefer.

ivmaykov replied:
Fair enough, I'll do some refactoring next week

anmolnar replied:
Thanks @ivmaykov .

ivmaykov replied:
@anmolnar actually I'll wait for the rest of your review so I can get through all the initial changes in one pass.

ivmaykov replied:
@anmolnar I believe this is done.

anmolnar commented on zookeeper-server/src/main/java/org/apache/zookeeper/util/PemReader.java:
```diff
@@ -0,0 +1,224 @@
+
+/**
+ * Note: this class is copied from io.airlift.security.pem.PemReader (see
+ * https://github.com/airlift/airlift/blob/master/security/src/main/java/io/airlift/security/pem/PemReader.java) with
```
You could be right, I'm not sure what's the implication of this. @hanm must be able to advise on this.
BouncyCastle library also having some PEM reader logic, but not sure how useful it is here.

ivmaykov replied:
BouncyCastle is currently a test-only dependency (used for creating test certificates), so we can't use it in production code. It's a large library and I'd rather not take it on as a full dependency if the only thing we need from it is the PEM reader. I think it's fine to copy the code (license allows it and the author of airlift recommended it to me), will wait for guidance on NOTICE file update from @eolivelli.

eolivelli replied:
For the NOTICE you can take a look to this example from Bookkeeper project
https://github.com/apache/bookkeeper/blob/master/bookkeeper-dist/src/main/resources/NOTICE-server.bin.txt

hanm replied:
i agree with @eolivelli , it should be enough to declare airlift in our notice file as a dependency.
our notice file is here:
https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/resources/NOTICE.txt

I vaguely remember we also copied some code from org.apache.httpcomponents for host verification to mitigate the perf regression of directly referencing the library as a dependency. Might worth to double check if we need patch notice file for that copy as well.

ivmaykov replied:
@hanm I've updated the NOTICE.txt file in the latest version of the code, please let me know if that looks good. I think it doesn't hurt to keep the comments in PemReader.java - will make it easier to backport fixes from airlift's PemReader in the future.

RE: httpcomponents. Yes, we did copy some httpcomponents code. I can put up a separate PR that adds a NOTICE.txt entry for that.

hanm replied:
notice file lgtm, thanks

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -167,47 +222,50 @@ public SSLContext createSSLContext(ZKConfig config) throws SSLContextException {
 
-        String keyStoreLocationProp = config.getProperty(sslKeystoreLocationProperty);
-        String keyStorePasswordProp = config.getProperty(sslKeystorePasswdProperty);
+        String keyStoreLocationProp = config.getProperty(sslKeystoreLocationProperty, "");
```
Since empty string is not a valid keystore type, I treat null value and empty string values equivalently. Doing it this way simplifies the check below (only need to check for empty string instead of checking both null and empty).

ivmaykov replied:
Sorry, I misread this comment. You are talking about the location, not the store type. But it's the same thing - null location and empty location can be treated the same way, so by defaulting to empty string here we don't need to check for 2 different options below. The previous code actually didn't check for empty location at all.

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -79,12 +82,56 @@
     private String cipherSuitesProperty = getConfigPrefix() + "ciphersuites";
     private String sslKeystoreLocationProperty = getConfigPrefix() + "keyStore.location";
     private String sslKeystorePasswdProperty = getConfigPrefix() + "keyStore.password";
+    private String sslKeystoreTypeProperty = getConfigPrefix() + "keyStore.type";
```
I think it makes sense, if someone puts their JKS key in a file that they name "foobar.key" or something non-standard, we can still support it if they explicitly set the store type option. Most people will probably use the default extensions and leave this option unset.

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -167,47 +222,50 @@ public SSLContext createSSLContext(ZKConfig config) throws SSLContextException {
 
-        String trustStoreLocationProp = config.getProperty(sslTruststoreLocationProperty);
-        String trustStorePasswordProp = config.getProperty(sslTruststorePasswdProperty);
+        String trustStoreLocationProp = config.getProperty(sslTruststoreLocationProperty, "");
```
See above

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -360,4 +476,26 @@ private void configureSSLServerSocket(SSLServerSocket sslServerSocket) {
+     * @return a StoreFileType.
+     * @throws IOException if the filename does not end with ".jks" or ".pem".
+     */
+    public static StoreFileType detectStoreFileTypeFromFileExtension(File filename) throws IOException {
```
How about I make it package-private and add a unit test?

ivmaykov commented:
@anmolnar added KeyStoreLoader classes
@eolivelli updated NOTICE file and PemReader.java

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2532/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2536/

eolivelli commented on NOTICE.txt:
```diff
@@ -3,3 +3,9 @@ Copyright 2009-2014 The Apache Software Foundation
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
+
+This product includes software components originally
```
Looks good.
Nit: Airlift with the uppercase A

ivmaykov commented:
@anmolnar it's certainly easier to keep track of the copied code if we don't modify it much. Moving the logic into PEMFileLoader and making it non-static would make it harder to trace the history.

ivmaykov commented:
@eolivelli capitalized "Airlift" in NOTICE.txt
@anmolnar use `FileNameUtils.getExtension()` for file type detection
everyone: fixed some copy-paste bugs in PEMFileLoaderTest and JKSFileLoaderTest, and added more test cases.

ivmaykov commented:
Refactored common code shared by a bunch of tests that use `X509TestContext` into a new base class, `BaseX509ParameterizedTestCase`.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2544/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2548/

ivmaykov commented:
Revert `FileNameUtils` change as it seems to be breaking contbuild

ivmaykov commented:
@eolivelli what does a "(non binding)" +1 mean? Can we merge this to upstream/master at this point?

eolivelli commented:
'Non binding' means that I am not a 'committer' so I can't merge the patch by myself.
Btw the other guys which are reviewing this patch are committers so I think this patch will be merged very soon.
Thanks

ivmaykov commented:
@eolivelli got it, thanks! I will use that terminology from now on :)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2555/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2559/

anmolnar commented:
@ivmaykov What was the problem with using `FileNameUtils`?

tumativ commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/FileKeyStoreLoader.java:
```diff
@@ -0,0 +1,98 @@
+     * @param type the file type to load keys/certs from.
+     * @return a new Builder.
+     */
+    static Builder<? extends FileKeyStoreLoader> getBuilderForKeyStoreFileType(
```
getBuilderForKeyStoreFileType is depending on loader implementation classes  .This may lead to the subclasses of this also depend on other loader implementation classes. How about moving this to something like BuilderProvider?

ivmaykov replied:
Okay :( This is the part of java I dislike. The good old "AbstractBaseFooFactoryFactory" pattern :/

Relevant: https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition

ivmaykov replied:
@tumativ this is done, please see the latest version of the code

tumativ commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/JKSFileLoader.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public KeyStore loadKeyStore() throws IOException, GeneralSecurityException {
+        KeyStore ks = KeyStore.getInstance("JKS");
```
Can we declare global final variable for "JKS"  and refer in code?

ivmaykov replied:
Will do

tumativ commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/FileKeyStoreLoader.java:
```diff
@@ -0,0 +1,98 @@
+    final String keyStorePassword;
+    final String trustStorePassword;
+
+    FileKeyStoreLoader(String keyStorePath,
```
How about changing the keyStorePath and trustStorePathto URL?So that it can be leveraged for other use cases?Let builder take file path and convert it to URL.

ivmaykov replied:
Should I rename the class to URLKeyStoreLoader then?

ivmaykov replied:
Actually, the PemReader class doesn't currently support `URL`, only `File`. So I'd prefer to only support file paths here for now. We can modify it later to support loading from URLs, in a separate PR / JIRA.

ivmaykov replied:
Also, loading private keys from arbitrary URLs goes against best security practices (unless that URL is for a local file, in which case there is no point in wrapping it in a URL). TLS private keys should never be transmitted across a network. The CA cert (i.e. Trust Store) could in theory be loaded from a remote URL though.

tumativ replied:
Yaa.Thanks for explaining

tumativ commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -254,23 +282,54 @@ public static X509KeyManager createKeyManager(String keyStoreLocation, String ke
-                ts.load(inputStream, null);
-            }
-
+            KeyStoreFileType storeFileType =
```
IllegalArgumentException is possible here. I see it not handled

ivmaykov replied:
Will fix.

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -221,27 +229,47 @@ public SSLContext createSSLContext(ZKConfig config) throws SSLContextException {
-            KeyStore ks = KeyStore.getInstance("JKS");
-            inputStream = new FileInputStream(keyStoreFile);
-            ks.load(inputStream, keyStorePasswordChars);
+            KeyStoreFileType storeFileType =
```
Will fix. This also made me realize that `JKSFileLoader` and `PEMFileLoader` are leaking file input streams. Will fix that as well.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2563/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2584/

anmolnar commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2589/

ivmaykov commented:
@anmolnar it works for me as far as I can tell, I think the contbuild is flaky? The contbuild on #681, which includes the same commit, passes.

ivmaykov commented:
@anmolnar is there a way to reproduce the exact same steps that jenkins runs on my macbook?

eolivelli commented:
Maybe you should use docker.
As soon as possible I will copy the command line we are using in Jenkins job.
Maybe it is already available on the job log

anmolnar commented:
@ivmaykov It's in:
`zookeeper-server/src/test/resources/test-github-pr.sh`

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2599/

ivmaykov commented:
I don't understand why this PR started failing Jenkins builds all of a sudden. The next 2 PRs stacked on this one (#678 and #680) also fail, but the last one (#681) passes.

ivmaykov commented:
Jenkins claims there is a findbugs failure, but there is no actual output of findbugs failures, and findbugs passes on my machine. ¯\_(ツ)_/¯

anmolnar commented:
It's not findbugs, core and contrib tests are failing, but I cannot see why:
```     
     [exec]     -1 core tests.  The patch failed core unit tests.
     [exec]     -1 contrib tests.  The patch failed contrib unit tests.
 ```

enixon commented:
Yeah, there aren't findbugs reports in the artifacts as I would expect https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2599/artifact/patchprocess/ , it's unclear what went wrong there. Maybe retrigger the build process ('retest this please')?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2604/

anmolnar commented:
Now it's green. I'll merge it.
Previously the findbugs subprocess has been killed for some reason.

anmolnar commented:
Merged to 3.5 and master branches. Thanks @ivmaykov !

ivmaykov commented:
@anmolnar thanks for merging this :) I've rebased #679, #680, and #681 on top of master. Let's get those in soon :)

sachmo99 commented:
Is this commit only for Quorum TLS or client-zookeeper TLS too? If it's only for Quorum TLS, can anyone please point out where I can find client-zookeeper TLS using PEM is detailed? The documentation mentions PEM is supported but I keep getting "DER input, Integer Tag ..." error. So sorry if I am posting this comment in the wrong space.  

_"ssl.keyStore.type and ssl.quorum.keyStore.type : (Java system properties: zookeeper.ssl.keyStore.type and zookeeper.ssl.quorum.keyStore.type) New in 3.5.5: Specifies the file format of client and quorum keystores. Values: JKS, PEM, PKCS12 or null (detect by filename).
Default: null"_

eolivelli commented:
@sachmo99 please send a message to user@zookeeper.apache.org

that way you will find more answers
https://zookeeper.apache.org/lists.html
